src/gui/touchscreengui.cpp

```diff
@@ -250,367 +327,236 @@ void AutoHideButtonBar::addButton(touch_gui_button_id button_id, const wchar_t *
-	btn->toggleable = button_info::FIRST_TEXTURE;
-	btn->textures.push_back(btn_image_1);
-	btn->textures.push_back(btn_image_2);
+	addButton(id, image_1);
+	button_info &btn = m_buttons.back();
+	btn.toggleable = button_info::FIRST_TEXTURE;
+	btn.toggle_textures[0] = image_1;
+	btn.toggle_textures[1] = image_2;
 }
 
-bool AutoHideButtonBar::isButton(const SEvent &event)
+bool AutoHideButtonBar::handlePress(size_t pointer_id, IGUIElement *element)
 {
-	IGUIElement *rootguielement = m_guienv->getRootGUIElement();
-
-	if (rootguielement == nullptr)
-		return false;
-
-	gui::IGUIElement *element = rootguielement->getElementFromPoint(
-			v2s32(event.TouchInput.X, event.TouchInput.Y));
-
-	if (element == nullptr)
-		return false;
-
 	if (m_active) {
-		// check for all buttons in vector
-		for (const auto &button : m_buttons) {
-			if (button->gui_button == element) {
-				SEvent translated{};
-				translated.EventType        = irr::EET_KEY_INPUT_EVENT;
-				translated.KeyInput.Key     = button->keycode;
-				translated.KeyInput.Control = false;
-				translated.KeyInput.Shift   = false;
-				translated.KeyInput.Char    = 0;
-
-				// add this event
-				translated.KeyInput.PressedDown = true;
-				m_receiver->OnEvent(translated);
-
-				// remove this event
-				translated.KeyInput.PressedDown = false;
-				m_receiver->OnEvent(translated);
-
-				button->ids.push_back(event.TouchInput.ID);
-
-				m_timeout = 0.0f;
-
-				if (button->toggleable == button_info::FIRST_TEXTURE) {
-					button->toggleable = button_info::SECOND_TEXTURE;
-					load_button_texture(button.get(), button->textures[1],
-							button->gui_button->getRelativePosition(),
-							m_texturesource, m_driver);
-				} else if (button->toggleable == button_info::SECOND_TEXTURE) {
-					button->toggleable = button_info::FIRST_TEXTURE;
-					load_button_texture(button.get(), button->textures[0],
-							button->gui_button->getRelativePosition(),
-							m_texturesource, m_driver);
-				}
-
-				return true;
-			}
-		}
-	} else {
-		// check for starter button only
-		if (element == m_starter.gui_button) {
-			m_starter.ids.push_back(event.TouchInput.ID);
-			m_starter.gui_button->setVisible(false);
-			m_starter.gui_button->setEnabled(false);
-			m_active = true;
-			m_timeout = 0.0f;
-
-			for (const auto &button : m_buttons) {
-				button->gui_button->setVisible(true);
-				button->gui_button->setEnabled(true);
-			}
-
-			return true;
-		}
+		return buttons_handlePress(m_buttons, pointer_id, element, m_driver,
+				m_receiver, m_texturesource);
+	}
+	if (m_starter.get() == element) {
+		activate();
+		return true;
 	}
 	return false;
 }
 
+bool AutoHideButtonBar::handleRelease(size_t pointer_id)
+{
+	return buttons_handleRelease(m_buttons, pointer_id, m_driver,
+			m_receiver, m_texturesource);
+}
+
 void AutoHideButtonBar::step(float dtime)
 {
 	if (m_active) {
-		m_timeout += dtime;
+		bool has_pointers = buttons_step(m_buttons, dtime, m_driver,
+				m_receiver, m_texturesource);
```

Comment:
always step in case someone opens another bar with a different finger while still keeping a button pressed?